File: adventofcode/2020/day7/day7.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

def parse_bag_color(bag)
  color = bag.split(/bag/).first.strip
  color = color[2..-1] if color =~ /^\d/
  color
end

def bag_rules(bag_data)
  rules = {}
  bag_data.each do |bag|
    outter_bag, inner_bags = bag.split(/contain/)
    outter_bag_color = parse_bag_color(outter_bag)
    inner_bag_colors = inner_bags.split(',').map { |b| parse_bag_color(b) }
    rules[outter_bag_color] = inner_bag_colors
  end
  rules
end

SHINY_BAG = 'shiny gold'
LEAF_BAG = 'no other'

def can_contain?(rules, big, small)
  return false unless rules[big]
  return true if rules[big].include?(small)

  rules[big].any? { |b| can_contain?(rules, b, small) }
end

def part1(bag_rules)
  bag_rules.keys.select do |color|
    can_contain?(bag_rules, color, SHINY_BAG)
  end.size
end

def parse_bag(bag)
  color = bag.split(/bag/).first.strip
  count = 1
  if color =~ /^\d/
    count = color.split.first.to_i
    color = color.split[1..-1].join(' ')
  end
  { color: color, count: count }
end

def bag_rules2(bag_data)
  rules = {}
  bag_data.each do |bag|
    outter_bag, inner_bags = bag.split(/contain/)
    outter_bag_color = parse_bag(outter_bag)
    if inner_bags.match(LEAF_BAG)
      rules[outter_bag_color[:color]] = nil
    else
      inner_bag_colors = inner_bags.split(',').map { |b| parse_bag(b) }
      rules[outter_bag_color[:color]] = inner_bag_colors
    end
  end
  rules
end

def contains(rules, big_color)
  return 0 unless rules[big_color]

  count = 0
  rules[big_color].each do |bag|
    count += (1 + contains(rules, bag[:color])) * bag[:count]
  end
  count
end

def part2(bag_rules)
  contains(bag_rules, SHINY_BAG)
end

bag_data = read_input
p part1(bag_rules(bag_data))
p part2(bag_rules2(bag_data))

File: adventofcode/2020/day7/day7.hs
import Data.Char
import qualified Data.Map as M
import qualified Data.Text as T

data Bag = Bag Int T.Text deriving (Show)

instance Eq Bag where
  (==) (Bag x xs) (Bag y ys) = x == y && xs == ys

instance Ord Bag where
  compare (Bag x xs) (Bag y ys)
    | x == y = compare xs ys
    | otherwise = compare x y

comma = T.pack ","
bagDelim = T.pack "bag"
conDelim = T.pack " contain "

parseBag :: T.Text -> Bag
parseBag bs
  | isDigit c = Bag (digitToInt c) rs
  | otherwise = Bag 0 bs
  where
    c = T.head bs
    rs = T.strip $ T.tail bs

bagRule :: T.Text -> [(T.Text, [Bag])]
bagRule ls = [(pclr, cbags)]
  where
    [pbs, cbs] = map T.strip $ T.splitOn conDelim ls
    (Bag _ pclr) = parseBag $ head $ map T.strip $ T.splitOn bagDelim pbs
    bagf = parseBag . T.strip . head . T.splitOn bagDelim
    cbags = map bagf $ T.splitOn comma cbs

bagRules :: [T.Text] -> M.Map T.Text [Bag]
bagRules = M.fromList . concatMap bagRule

canContains :: M.Map T.Text [Bag] -> T.Text -> T.Text -> Bool
canContains bmap sclr bclr
  | M.notMember bclr bmap = False
  | any (\(Bag _ clr) -> sclr == clr) bags = True
  | otherwise = any (\(Bag _ clr) -> canContains bmap sclr clr) bags
  where (Just bags) = M.lookup bclr bmap

shinyBag = T.pack "shiny gold"
brightWhiteBag = T.pack "bright white"
darkOrangeBag = T.pack "dark orange"

part1 :: [T.Text] -> Int
part1 ls = length ks
  where bmap = bagRules ls
        ks = filter (canContains bmap shinyBag) $ M.keys bmap

contains :: M.Map T.Text [Bag] -> T.Text -> Int
contains bmap bclr
  | M.notMember bclr bmap = 0
  | otherwise = foldl f 0 bags
  where (Just bags) = M.lookup bclr bmap
        f r (Bag c clr) = r + (1 + contains bmap clr) * c

part2 :: [T.Text] -> Int
part2 ls = contains bmap shinyBag
  where bmap = bagRules ls

solve :: [T.Text] -> String
solve ls = unlines $ map show [part1 ls, part2 ls]

main :: IO ()
main = interact $ solve . T.lines . T.pack
